stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+
+  g.w[0] = twopn * ((12824 - 9120 * n + 400 * n * n) * 1.0 / 19683);
+  g.w[1] = twopn * (980.0 / 6561);
+  g.w[2] = twopn * ((1820 - 400 * n) * 1.0 / 19683);
+  g.w[3] = twopn * (200.0 / 19683);
+  g.w[4] = 6859.0 / 19683;
+  g.wd[3] = twopn * (25.0 / 729);
+  g.wd[2] = twopn * ((265 - 100 * n) * 1.0 / 1458);
+  g.wd[1] = twopn * (245.0 / 486);
+  g.wd[0] = twopn * ((729 - 950 * n + 50 * n * n) * 1.0 / 729);
+
+  combos(1, l2, n, g.p[0]);
+  combos(1, l3, n, g.p[1]);
+  signcombos(2, l4, n, g.p[2]);
+  signcombos(n, l5, n, g.p[3]);
+}
+
+template <typename T_GenzMalik>
+void clean_GenzMalik(T_GenzMalik& g) {
+  for (size_t j = 0; j != 4; j++)
+    for (size_t i = 0; i != g.p[j].size(); i++)
+      if (g.p[j][i])
+        free(g.p[j][i]);
+}
+
+template <typename F, typename T_GenzMalik, typename T_n, typename T_a,
+          typename T_b, typename T_oned, typename T_pars>
+void integrate_GenzMalik(const F& integrand, T_GenzMalik& g, const T_n& n,
+                         const T_a& a, const T_b& b, T_oned& out,
+                         T_pars& pars) {
+  using T_return_type = return_type_t<T_n, T_a, T_b>;
+  std::vector<T_return_type> c(n, 0);
+  double* deltac = reinterpret_cast<double*>(malloc(n * sizeof(double)));
+
+  for (size_t i = 0; i != n; i++)
+    c[i] = (a[i] + b[i]) / 2;
+
+  for (size_t i = 0; i != n; i++)
+    deltac[i] = fabs(b[i] - a[i]) / 2;
+  T_return_type v = 1.0;
+  for (size_t i = 0; i != n; i++)
+    v *= deltac[i];
+
+  if (v == 0.0) {
+    out.err = 0.0;
+    out.result = 0.0;
+    out.kdivide = 0;
+    free(deltac);
+    return;
+  }
+
+  T_return_type f1 = integrand(c, pars);
+  T_return_type f2 = 0.0;
+  T_return_type f3 = 0.0;
+  T_return_type twelvef1 = 12 * f1;
+
+  T_return_type maxdivdiff = 0.0;
+  T_return_type* divdiff
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
+  T_return_type* p2
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
+  T_return_type* p3
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
+  std::vector<T_return_type> cc(n, 0);
+
+  for (size_t i = 0; i != n; i++) {
+    for (size_t j = 0; j != n; j++)
+      p2[j] = deltac[j] * g.p[0][i][j];
+
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p2[j];
+    T_return_type f2i = integrand(cc, pars);
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] - p2[j];
+    T_return_type temp = integrand(cc, pars);
+    f2i += temp;
+
+    for (size_t j = 0; j != n; j++)
+      p3[j] = deltac[j] * g.p[1][i][j];
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p3[j];
+    T_return_type f3i = integrand(cc, pars);
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] - p3[j];
+    temp = integrand(cc, pars);
+    f3i += temp;
+    f2 += f2i;
+    f3 += f3i;
+    divdiff[i] = fabs(f3i + twelvef1 - 7 * f2i);
+  }
+  free(p2);
+  free(p3);
+  T_return_type* p4
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
```

Comment:
I think where you are doing raw mallocs now you can use something like 

```suggestion
  Eigen::Array<T_return_type, -1, 1> p4(n);
```

Reply:
As mentioned above, all mallocs are changed into vectors.
Comes with next commit.